lib/omnibus/packagers/deb.rb

```diff
@@ -390,6 +415,73 @@ def compression_params
     end
 
     #
+    # Check for the existance of a command in +ENV['PATH']+. Exit code +127+
+    # tells us the shell could not find the program.
+    #
+    # @return [bool]
+    #
+    def check_command_exists(command)
+      !(shellout("#{command} </dev/null >/dev/null 2>&1").exitstatus == 127)
+    end
+
+    #
+    # Sign the  +.deb+ file with gpg. This has to be done as separate steps
+    # from creating the +.deb+ file. See +debsigs+ source for behavior
+    # replicated here. +https://gitlab.com/debsigs/debsigs/blob/master/debsigs.txt#L103-124+
+    #
+    # @return [void]
+    def sign_deb_file
+      if signing_passphrase
+        log.info(log_key) { "Signing enabled for .deb file" }
+
+        # Check our dependencies and determine command for GnuPG
+        gpg = nil
+        if check_command_exists('gpg2')
+          gpg = 'gpg2'
+        elsif check_command_exists('gpg')
+          gpg = 'gpg'
+        end
+
+        if gpg && check_command_exists('ar')
+          # Create a directory that will be cleaned when we leave the block
+          Dir.mktmpdir { |tmp_dir|
+            Dir.chdir(tmp_dir) do
+              # Extract the deb file contents
+              shellout!("ar x #{Config.package_dir}/#{package_name}")
+              # Concantenate contents, in order per +debsigs+ documentation.
+              shellout!("cat debian-binary control.tar.* data.tar.* > complete")
+              # Create signature (as +root+)
+              gpg_command =  "#{gpg} --armor --sign --detach-sign"
+              gpg_command << " --local-user '#{project.maintainer}'"
+              gpg_command << " --homedir #{ENV['HOME']}/.gnupg" # TODO: Make this configurable
+              ## pass the +signing_passphrase+ via +STDIN+
+              gpg_command << " --batch --no-tty"
+              ## Check `gpg` for the compatibility/need of pinentry-mode
+              # - We're calling gpg with the +--pinentry-mode+ argument, and +STDIN+ of +/dev/null+
+              # - This _will_ fail with exit code 2 no matter what. We want to check the +STDERR+
+              #   for the error message about the parameter. If it is _not present_ in the
+              #   output, then we _do_ want to add it. (If +grep -q+ is +1+, add parameter)
+              if shellout("#{gpg} --pinentry-mode loopback </dev/null 2>&1 | grep -q pinentry-mode").exitstatus == 1
+                gpg_command << " --pinentry-mode loopback"
+              end
+              gpg_command << " --passphrase-fd 0"
+              gpg_command << " -o _gpgorigin complete"
+              shellout!("fakeroot #{gpg_command}",
+                input: signing_passphrase
+              )
+              # Append +_gpgorigin+ to the +.deb+ file (as +root+)
+              shellout!("fakeroot ar rc #{Config.package_dir}/#{package_name} _gpgorigin")
```

Comment:
The `fakeroot` binary might not be installed on some systems, thus to be overly cautious you could also check for its existence - similarly to `gpg`/`gpg2` and `ar`.

Reply:
we're actually guaranteed that `fakeroot` is present in the omnibus cookbook here: https://github.com/chef-cookbooks/omnibus/blob/master/recipes/_packaging.rb#L32

Reply:
Indeed this caused test failures above. I'll have to recheck my logic on why I did this with `fakeroot`.

Reply:
It appears I did this so that the archive (deb file) behaves as expected. Considering best practices for deb files to have their contents owned by root. It is also used by the actual `create_deb_file` function that is already a part of the package.

Reply:
@scotthain ah, OK. Albeit, not everyone would use the cookbook. I personally don't use it in our internal build system (at work at the moment). Which is where the idea came from.

Nonetheless, should there be a parallel Pull Request against the cookbook to install GnuPG and `ar`? What do you think, @scotthain and @WarheadsSE?

Reply:
@kwilczynski Yeah a PR in the omnibus cookbook would be fantastic, just for safety's sake. Currently they are part of the base install of the systems we use internally at Chef but that could easily change in the future.

Reply:
@scotthain Done. Pull Request opened at https://github.com/chef-cookbooks/omnibus/pull/217.